File: golang/2364_count_bad_pairs/count_bad_pairs.go
// Package main ...
package main

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func countBadPairs(nums []int) int64 {
	mp := make(map[int]int)
	var res int64
	for i := 0; i < len(nums); i++ {
		key := nums[i] - i
		res += int64(i - mp[key])
		mp[key]++
	}
	return res
}

func main() {
	tests := []struct {
		nums []int
		ans  int64
	}{
		{[]int{4, 1, 3, 3}, 5},
		{[]int{1, 2, 3, 4, 5}, 0},
	}

	for index, test := range tests {
		assert.Equal(&testing.T{}, test.ans, countBadPairs(test.nums), index)
	}
}

File: cpp/2364_count_bad_pairs/count_bad_pairs.cpp
#include <cassert>
#include <tuple>
#include <unordered_map>
#include <vector>

using namespace std;

class Solution {
 public:
  using ll = long long;
  long long countBadPairs(vector<int>& nums) {
    unordered_map<int, int> mp;
    ll res = 0;
    for (int i = 0; i < nums.size(); i++) {
      res += i - mp[nums[i] - i];
      mp[nums[i] - i]++;
    }
    return res;
  }
};

int main() {
  vector<tuple<vector<int>, long long>> tests{
      {{4, 1, 3, 3}, 5},
      {{1, 2, 3, 4, 5}, 0},
  };

  for (auto& [nums, ans] : tests) {
    assert(Solution().countBadPairs(nums) == ans);
  }
}